consensus/src/pending_votes.rs

```diff
@@ -56,15 +62,117 @@ pub enum VoteReceptionResult {
     EchoTimeout(u128),
 }
 
+#[derive(Debug)]
+pub(super) struct TwoChainTimeoutVotes {
+    timeout_reason: HashMap<Author, RoundTimeoutReason>,
+    partial_2chain_tc: TwoChainTimeoutWithPartialSignatures,
+}
+
+impl TwoChainTimeoutVotes {
+    pub(super) fn new(timeout: TwoChainTimeout) -> Self {
+        Self {
+            partial_2chain_tc: TwoChainTimeoutWithPartialSignatures::new(timeout.clone()),
+            timeout_reason: HashMap::new(),
+        }
+    }
+
+    pub(super) fn add(
+        &mut self,
+        author: Author,
+        timeout: TwoChainTimeout,
+        signature: bls12381::Signature,
+        reason: RoundTimeoutReason,
+    ) {
+        self.partial_2chain_tc.add(author, timeout, signature);
+        self.timeout_reason.entry(author).or_insert(reason);
+    }
+
+    pub(super) fn partial_2chain_tc_mut(&mut self) -> &mut TwoChainTimeoutWithPartialSignatures {
+        &mut self.partial_2chain_tc
+    }
+
+    pub(super) fn partial_2chain_tc(&self) -> &TwoChainTimeoutWithPartialSignatures {
+        &self.partial_2chain_tc
+    }
+
+    fn aggregated_timeout_reason(&self, verifier: &ValidatorVerifier) -> RoundTimeoutReason {
+        let mut reason_voting_power: HashMap<RoundTimeoutReason, u128> = HashMap::new();
+        let mut missing_batch_authors: HashMap<usize, u128> = HashMap::new();
+        // let ordered_authors = verifier.get_ordered_account_addresses();
+        for (author, reason) in &self.timeout_reason {
+            // To aggregate the reason, we only care about the variant type itself and
+            // exclude any data within the variants.
+            let reason_key = match reason {
+                reason @ RoundTimeoutReason::Unknown
+                | reason @ RoundTimeoutReason::ProposalNotReceived
+                | reason @ RoundTimeoutReason::NoQC => reason.clone(),
+                RoundTimeoutReason::PayloadUnavailable { missing_authors } => {
+                    for missing_idx in missing_authors.iter_ones() {
+                        *missing_batch_authors.entry(missing_idx).or_default() +=
+                            verifier.get_voting_power(author).unwrap() as u128;
+                    }
+                    RoundTimeoutReason::PayloadUnavailable {
+                        missing_authors: BitVec::with_num_bits(verifier.len() as u16),
+                    }
+                },
+            };
+            *reason_voting_power.entry(reason_key).or_default() +=
+                verifier.get_voting_power(&author).unwrap_or_default() as u128;
+        }
+        // The aggregated timeout reason is the reason with the most voting power received from
+        // at least f+1 peers by voting power. If such voting power does not exist, then the
```

Comment:
I wonder if the f+1 requirement for marking a malicious batch proposer is too strong or not. Ofc it is safer against malicious nodes trying to make honest nodes malicious, but anyway malicious nodes can disable optQS?

Reply:
f+1 is straight-forward, anything less, we need to collect data and re-calibrate. I don't want it to be a fixed number and be tied to f somehow so it can scale with growing network.

Reply:
I am just concerned that slow nodes are not marked malicious under the f+1 requirement. But we can leave it for now and see how it performs.